cloud-gov/cg-ui
# #494: If a user is unauthenticated and attempts to visit an authenticated page, redirect them to UAA

## Changes proposed in this pull request:

- Redirect to login page when going to an authenticated route
- Redirect to login page when trying to perform an action on a page and token is expired
- After logging in, redirect user to the page they were originally trying to get it
- And while we're at it, let's convert the auth functionality to Typescript 

### Notes

We need to authenticate the user in a few scenarios:
1. When requesting an authenticated route, and
2. When making a CloudFoundry API request

(Sidebar on why we need to authenticate in scenario 2: Let’s say a user has the Manage Users page open. They leave it open for a while and eventually, their auth token expires, making this user no longer authenticated. Then they go back to this page and try to edit a member’s org roles by opening an overlay. This overlay is not a new URL route, but it does makes an API request to show the org form. In this case, if the user is unauthenticated, then we should kick them back to the login page if they try to perform this action.) 

Scenario 1 is triggered through Middleware, which checks the route and routes them to /login if unauthenticated.

Scenario 2 is triggered in `cfRequest`, which all CF API requests use. 

Known bug: because the current url path is [only obtainable in client-side code](https://nextjs.org/docs/app/api-reference/functions/use-pathname) or in middleware, I’m not able to set the last desired path for scenario 2, only scenario 1. 

### How to test

In order to keep the developer experience lightweight, I'm bypassing authentication in development environments, so you can either:
1. push this code to our deployed site and test there, or
2. remove the `process.env.NODE_ENV === 'development'` checks in  `authenticateRoute` and `cfRequest`

Steps:
1. Be logged out of the site (i.e. no 'authsession' cookie should be present in your browser)
2. Try to access an authenticated route (`/orgs` is a good one`)
3. Go through the UAA login routine
4. Be redirected to your original desired page (like `/orgs`)

Steps for scenario 2:
1. Log in an go to a Manage Users page for an organization
2. Open your dev tools and clear your `authsession` cookie
3. Try to open an edit org roles overlay
4. Be redirected to login page

### Related issues

Closes #494 

### Submitter checklist


## Security considerations

Improves security by ensuring unauthorized users do not access pages they shouldn't
